# Week 7: Comparing Multiple Machine Learning Models

## One Model Isn't Enough

In Week 6, I trained and evaluated my first machine learning model. In this task, I will train three different classification models and compare their performance fairly.

The three models are:

1. K-Nearest Neighbors
2. Decision Tree
3. Logistic Regression

All three models will use the same training and testing data.

## Project Question

Which beginner classification model performs best when predicting the species of an Iris flower using its sepal and petal measurements?

This is a classification problem because the model predicts one of three flower species:

- Setosa
- Versicolor
- Virginica

## Learning Notes: Why One Model Is Not Enough

No single machine learning algorithm is best for every problem. Different algorithms learn patterns in different ways.

This means that I should not select a model only because it is popular or because it worked well on another dataset. I need to train different models and compare their performance on the same data.

This idea is sometimes called the "no free lunch" principle. In simple words, it means that every model has strengths and weaknesses, and the best model depends on the problem and the dataset.

## How the Three Models Think

### K-Nearest Neighbors

KNN predicts a new answer by checking the closest examples in the training data. If most nearby flowers belong to the same species, the model predicts that species.

### Decision Tree

A Decision Tree makes predictions by asking a series of yes-or-no questions about the data. For example, it may check whether the petal length is smaller or greater than a certain value.

### Logistic Regression

Logistic Regression learns boundaries between different classes. It uses the flower measurements to decide which side of the learned boundary a new flower belongs to.

## Fair Model Comparison

For a fair comparison, every model must use the same training data and the same testing data.

If each model receives a different train/test split, one model may receive easier test examples than another. Therefore, I will create the train/test split only once and reuse it for all three models.

## Baseline

A baseline is a simple score that every real machine learning model should beat.

For this project, the baseline model will always predict the most common flower species. If a trained model cannot perform better than this simple method, the trained model is not very useful.

In [1]:
# Import the required libraries

import pandas as pd

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

In [2]:
# Load the Iris dataset

iris = load_iris()

print("The Iris dataset has been loaded successfully.")

The Iris dataset has been loaded successfully.


In [3]:
# Create a pandas DataFrame

df = pd.DataFrame(iris.data, columns=iris.feature_names)

# Add the target values
df["target"] = iris.target

# Add the readable flower species names
df["species"] = df["target"].map({
    0: "Setosa",
    1: "Versicolor",
    2: "Virginica"
})

df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,species
0,5.1,3.5,1.4,0.2,0,Setosa
1,4.9,3.0,1.4,0.2,0,Setosa
2,4.7,3.2,1.3,0.2,0,Setosa
3,4.6,3.1,1.5,0.2,0,Setosa
4,5.0,3.6,1.4,0.2,0,Setosa


## Dataset Information

The Iris dataset contains measurements from 150 flowers.

The input features are:

- Sepal length
- Sepal width
- Petal length
- Petal width

The target is the Iris flower species.

In [4]:
# Check the size of the dataset

print("Number of rows and columns:", df.shape)

# Check how many flowers belong to each species

print("\nNumber of flowers in each species:")
print(df["species"].value_counts())

Number of rows and columns: (150, 6)

Number of flowers in each species:
species
Setosa        50
Versicolor    50
Virginica     50
Name: count, dtype: int64


In [5]:
# Check for missing values

print("Missing values in each column:")
print(df.isnull().sum())

Missing values in each column:
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
species              0
dtype: int64


## Data Check

The dataset contains 150 flower records and has no missing values. Each flower species has 50 examples, so the classes are balanced.

Because the dataset is already clean, I can continue to the model preparation stage.

In [6]:
# Create X and y

X = df[iris.feature_names]
y = df["target"]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (150, 4)
y shape: (150,)


## Understanding X and y

X contains the input information used by the models. In this project, X contains the sepal and petal measurements.

y contains the correct answer that the models must learn to predict. In this project, y contains the Iris flower species number.

In [7]:
# Create one train/test split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training input shape:", X_train.shape)
print("Testing input shape:", X_test.shape)
print("Training target shape:", y_train.shape)
print("Testing target shape:", y_test.shape)

Training input shape: (120, 4)
Testing input shape: (30, 4)
Training target shape: (120,)
Testing target shape: (30,)


## Train/Test Split

I used 80% of the dataset for training and 20% for testing.

The random state was set to 42 so that the same split can be reproduced each time the notebook is run.

I also used stratify=y so that the training and testing sets contain a balanced number of examples from each flower species.

This split will be reused for all three models.

In [8]:
# Create and train the baseline model

baseline_model = DummyClassifier(strategy="most_frequent")

baseline_model.fit(X_train, y_train)

baseline_predictions = baseline_model.predict(X_test)

baseline_accuracy = accuracy_score(y_test, baseline_predictions)

print(f"Baseline Accuracy: {baseline_accuracy:.2%}")

Baseline Accuracy: 33.33%


## Baseline Observation

The baseline model achieved approximately 33.33% accuracy because it always predicted one flower species.

The three real machine learning models should perform better than this baseline.

In [9]:
# Create the K-Nearest Neighbors model

knn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model
knn_model.fit(X_train, y_train)

# Make predictions
knn_predictions = knn_model.predict(X_test)

# Calculate accuracy
knn_accuracy = accuracy_score(y_test, knn_predictions)

print(f"K-Nearest Neighbors Accuracy: {knn_accuracy:.2%}")

K-Nearest Neighbors Accuracy: 100.00%


## KNN Observation

The K-Nearest Neighbors model made predictions by checking the closest flower examples in the training data.

It performed well because flowers from the same species usually have similar sepal and petal measurements.

In [10]:
# Create the Decision Tree model

decision_tree_model = DecisionTreeClassifier(random_state=42)

# Train the model
decision_tree_model.fit(X_train, y_train)

# Make predictions
decision_tree_predictions = decision_tree_model.predict(X_test)

# Calculate accuracy
decision_tree_accuracy = accuracy_score(
    y_test,
    decision_tree_predictions
)

print(f"Decision Tree Accuracy: {decision_tree_accuracy:.2%}")

Decision Tree Accuracy: 93.33%


## Decision Tree Observation

The Decision Tree model learned a series of rules based on the flower measurements.

For example, it may use petal length and petal width to separate one flower species from another.

In [11]:
# Create the Logistic Regression model

logistic_regression_model = LogisticRegression(max_iter=200)

# Train the model
logistic_regression_model.fit(X_train, y_train)

# Make predictions
logistic_regression_predictions = logistic_regression_model.predict(
    X_test
)

# Calculate accuracy
logistic_regression_accuracy = accuracy_score(
    y_test,
    logistic_regression_predictions
)

print(
    f"Logistic Regression Accuracy: "
    f"{logistic_regression_accuracy:.2%}"
)

Logistic Regression Accuracy: 96.67%


## Logistic Regression Observation

The Logistic Regression model learned boundaries between the three flower species using the sepal and petal measurements.

It performed well because the measurements of the flower species can be separated reasonably clearly.

In [12]:
# Create the model comparison table

results = pd.DataFrame({
    "Model": [
        "K-Nearest Neighbors",
        "Decision Tree",
        "Logistic Regression"
    ],
    "Accuracy": [
        knn_accuracy,
        decision_tree_accuracy,
        logistic_regression_accuracy
    ]
})

# Arrange the models from highest to lowest accuracy

results = results.sort_values(
    by="Accuracy",
    ascending=False
).reset_index(drop=True)

# Create a percentage column

results["Accuracy (%)"] = (
    results["Accuracy"] * 100
).round(2)

# Display the final comparison table

results[["Model", "Accuracy (%)"]]

,Model,Accuracy (%)
0,K-Nearest Neighbors,100.00
1,Logistic Regression,96.67
2,Decision Tree,93.33


In [13]:
# Find the winning model and its margin over the second-best model

winning_model = results.iloc[0]
second_best_model = results.iloc[1]

winning_margin = (
    winning_model["Accuracy"] -
    second_best_model["Accuracy"]
) * 100

print("Winning model:", winning_model["Model"])
print(f"Winning accuracy: {winning_model['Accuracy']:.2%}")
print("Second-best model:", second_best_model["Model"])
print(f"Winning margin: {winning_margin:.2f} percentage points")

Winning model: K-Nearest Neighbors
Winning accuracy: 100.00%
Second-best model: Logistic Regression
Winning margin: 3.33 percentage points


## Written Verdict

The K-Nearest Neighbors model achieved the highest accuracy with 100.00%.

Logistic Regression was the second-best model with 96.67% accuracy. Therefore, KNN performed 3.33 percentage points better than Logistic Regression.

The Decision Tree achieved 93.33% accuracy.

My guess is that KNN suited the Iris dataset because flowers from the same species normally have similar sepal and petal measurements. This allowed KNN to make predictions by finding nearby flowers with similar measurements.

All three models performed better than the baseline accuracy of 33.33%. However, the test set contained only 30 flowers, so I should not assume that KNN will always be perfect on every new dataset.

## Final Conclusion

In this task, I trained and compared K-Nearest Neighbors, Decision Tree, and Logistic Regression using the same train/test split.

The comparison showed that different algorithms can produce different results even when they use the same dataset. KNN achieved the best result for this particular split, but this does not mean that KNN is the best model for every machine learning problem.

This task helped me understand why multiple models should be tested before selecting a final model.